packages/astro/src/core/session.ts

```diff
@@ -0,0 +1,406 @@
+	 */
+
+	destroy() {
+		this.#destroySafe();
+	}
+
+	/**
+	 * Regenerates the session, creating a new session ID. The existing session data is preserved.
+	 */
+
+	async regenerate() {
+		let data = new Map();
+		try {
+			data = await this.#ensureData();
+		} catch (err) {
+			// Log the error but continue with empty data
+			console.error('Failed to load session data during regeneration:', err);
+		}
+
+		// Store the old session ID for cleanup
+		const oldSessionId = this.#sessionID;
+
+		// Create new session
+		this.#sessionID = undefined;
+		this.#data = data;
+		this.#ensureSessionID();
+		await this.#setCookie();
+
+		// Clean up old session asynchronously
+		if (oldSessionId && this.#storage) {
+			this.#storage.removeItem(oldSessionId).catch((err) => {
+				console.error('Failed to remove old session data:', err);
+			});
+		}
+	}
+
+	// Persists the session data to storage.
+	// This is called automatically at the end of the request.
+	// Uses a symbol to prevent users from calling it directly.
+	async [PERSIST_SYMBOL]() {
+		// Handle session data persistence
+		if (this.#dirty && this.#data) {
+			const data = await this.#ensureData();
+			const key = this.#ensureSessionID();
+			let serialized;
+			try {
+				serialized = stringify(data, {
+					// Support URL objects
+					URL: (val) => val instanceof URL && val.href,
+				});
+			} catch (err) {
+				throw new AstroError(
+					{
+						...SessionStorageSaveError,
+						message: SessionStorageSaveError.message(
+							'The session data could not be serialized.',
+							this.#config.driver,
+						),
+					},
+					{ cause: err },
+				);
+			}
+			await this.#storage?.setItem(key, serialized);
+			this.#dirty = false;
+		}
+
+		// Handle destroyed session cleanup
+		if (this.#toDestroy.size > 0) {
+			await this.#ensureStorage();
+			const cleanupPromises = [...this.#toDestroy].map((sessionId) =>
+				this.#storage?.removeItem(sessionId).catch((err) => {
+					console.error(`Failed to clean up session ${sessionId}:`, err);
+				}),
+			);
+			await Promise.all(cleanupPromises);
+			this.#toDestroy.clear();
+		}
+	}
+
+	get sessionID() {
+		return this.#sessionID;
+	}
+
+	/**
+	 * Sets the session cookie.
+	 */
+	async #setCookie() {
+		if (!/^[\w-]+$/.test(this.#cookieName)) {
+			throw new AstroError({
+				...SessionStorageSaveError,
+				message: 'Invalid cookie name. Cookie names can only contain letters, numbers, and dashes.',
+			});
+		}
+		const cookieOptions: AstroCookieSetOptions = {
+			sameSite: 'lax',
+			secure: true,
+			path: '/',
+			...this.#cookieConfig,
+			httpOnly: true,
+		};
+		const value = this.#ensureSessionID();
+		this.#cookies.set(this.#cookieName, value, cookieOptions);
+	}
+
+	/**
+	 * Attempts to load the session data from storage, or creates a new data object if none exists.
+	 * If there is existing sparse data, it will be merged into the new data object.
+	 */
+
+	async #ensureData() {
+		await this.#ensureStorage();
```

Comment:
I know that typescript is weird, but it's also weird to see `ensureStorage` and then see `storage?.`, with the use of the chaining operator. Should we create a type guard, so that `this.#storage` isn't `undefined`? Or is there some other alternative that we can use? I'm worried that if we forget to call `this.#ensureStorage()` somewhere in the future, it will lead to unexpected errors.